File: ios/Classes/FlutterBraintreePlugin.swift
import Flutter
import UIKit
import Braintree


public class FlutterBraintreePlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_braintree_plugin", binaryMessenger: registrar.messenger())
    let instance = FlutterBraintreePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let authorization = FlutterBraintreePluginHelper.getAuthorization(call: call) else {
      FlutterBraintreePluginHelper.returnAuthorizationMissingError(result: result)
      return
    }
    
    guard let apiClient = BTAPIClient(authorization: authorization) else {
      FlutterBraintreePluginHelper.returnAuthorizationMissingError(result: result)
      return
    }
    
    switch call.method {
    case "tokenizePayPalAccount":
      tokenizePayPalAccount(call, apiClient: apiClient, result: result)
    case "tokenizeVenmoAccount":
      tokenizeVenmoAccount(call, apiClient: apiClient, result: result)
    case "fetchPaymentMethodNonces":
      let defaultFirst = FlutterBraintreePluginHelper.bool(for: "defaultFirst", in: call) ?? true
      fetchPaymentMethodNonces(apiClient: apiClient, defaultFirst: defaultFirst, result: result)
    case "deletePaymentMethodNonce":
      guard let nonce = FlutterBraintreePluginHelper.string(for: "payment_nonce", in: call) else {
        result([
          "error": [
            "code": "-1",
            "message": "Payment method does not exist",
          ]
        ])
        return
      }
      deletePaymentMethodNonce(apiClient: apiClient, nonce: nonce, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
  
  
  private func tokenizePayPalAccount(_ call: FlutterMethodCall, apiClient: BTAPIClient, result: @escaping FlutterResult) {
    
    guard let requestInfo = FlutterBraintreePluginHelper.dict(for: "request", in: call) else {
      FlutterBraintreePluginHelper.returnFlutterError(result: result, code: "-1", message: "Missing request parameters")
      return
    }
    
    guard let vault = requestInfo["vault"] as? Bool else {
      FlutterBraintreePluginHelper.returnFlutterError(result: result, code: "-1", message: "Missing request parameters")
      return
    }
    
    var paypalRequest:BTPayPalRequest;
    if vault {
      let paypalVaultRequest = BTPayPalVaultRequest()
      if let offerCredit = requestInfo["offerCredit"] as? Bool {
        paypalVaultRequest.offerCredit = offerCredit
      }
      paypalRequest = paypalVaultRequest
    } else {
      let amount = requestInfo["amount"] as? String ?? "0"
      let paypalCheckoutRequest = BTPayPalCheckoutRequest(amount: amount)
      paypalCheckoutRequest.currencyCode = requestInfo["currencyCode"] as? String
      if let intent = requestInfo["intent"] as? Int {
        switch intent {
        case 1:
          paypalCheckoutRequest.intent = BTPayPalRequestIntent.sale
        case 2:
          paypalCheckoutRequest.intent = BTPayPalRequestIntent.order
        default:
          paypalCheckoutRequest.intent = BTPayPalRequestIntent.authorize
        }
      }
      if let userAction = requestInfo["userAction"] as? Int {
        switch userAction {
        case 1:
          paypalCheckoutRequest.userAction = BTPayPalRequestUserAction.payNow
        default:
          paypalCheckoutRequest.userAction = BTPayPalRequestUserAction.none
        }
      }
      if let requestBillingAgreement = requestInfo["requestBillingAgreement"] as? Bool {
        paypalCheckoutRequest.requestBillingAgreement = requestBillingAgreement
      }
      if let offerPayLater = requestInfo["offerPayLater"] as? Bool {
        paypalCheckoutRequest.offerPayLater = offerPayLater
      }
      paypalRequest = paypalCheckoutRequest
    }
    
    paypalRequest.displayName = requestInfo["displayName"] as? String
    if let shippingAddressRequired = requestInfo["shippingAddressRequired"] as? Bool {
      paypalRequest.isShippingAddressRequired = shippingAddressRequired
    }
    if let billingAgreementDescription = requestInfo["billingAgreementDescription"] as? String {
      paypalRequest.billingAgreementDescription = billingAgreementDescription
    }
    if let shippingAddressOverride = requestInfo["shippingAddressOverride"] as? [String:Any] {
      let address = BTPostalAddress()
      if let recipientName = shippingAddressOverride["recipientName"] as? String {
        address.recipientName = recipientName
      }
      if let postalCode = shippingAddressOverride["postalCode"] as? String {
        address.postalCode = postalCode
      }
      if let streetAddress = shippingAddressOverride["streetAddress"] as? String {
        address.streetAddress = streetAddress
      }
      if let extendedAddress = shippingAddressOverride["extendedAddress"] as? String {
        address.extendedAddress = extendedAddress
      }
      if let locality = shippingAddressOverride["locality"] as? String {
        address.locality = locality
      }
      if let countryCodeAlpha2 = shippingAddressOverride["countryCodeAlpha2"] as? String {
        address.countryCodeAlpha2 = countryCodeAlpha2
      }
      if let region = shippingAddressOverride["region"] as? String {
        address.region = region
      }
      paypalRequest.shippingAddressOverride = address
    }
    if let shippingAddressEditable = requestInfo["shippingAddressEditable"] as? Bool {
      paypalRequest.isShippingAddressEditable = shippingAddressEditable
    }
    if let localeCode = requestInfo["localeCode"] as? Int {
      paypalRequest.localeCode = BTPayPalLocaleCode(rawValue: localeCode) ?? .none
    }
    if let merchantAccountID = requestInfo["merchantAccountID"] as? String {
      paypalRequest.merchantAccountID = merchantAccountID
    }
    if let riskCorrelationId = requestInfo["riskCorrelationId"] as? String {
      paypalRequest.riskCorrelationID = riskCorrelationId
    }
    if let landingPageType = requestInfo["landingPageType"] as? Int {
      switch landingPageType {
      case 1:
        paypalRequest.landingPageType = BTPayPalRequestLandingPageType.login
      case 2:
        paypalRequest.landingPageType = BTPayPalRequestLandingPageType.billing
      default:
        paypalRequest.landingPageType = BTPayPalRequestLandingPageType.none
      }
    }
    
    if let lineItems = requestInfo["lineItems"] as? [Any] {
      paypalRequest.lineItems = FlutterBraintreePluginHelper.makePayPalItems(from: lineItems)
    }
    
    let payPalClient = BTPayPalClient(apiClient: apiClient)
    
    if vault {
      payPalClient.tokenize(paypalRequest as! BTPayPalVaultRequest) { nonce, err in
        guard err == nil else {
          result([
            "error": [
              "message": String(describing: (err as? NSError)?.localizedDescription),
              "code": String(describing: (err as? NSError)?.code),
            ]
          ])
          return
        }
        result([
          "data": FlutterBraintreePluginHelper.buildPaymentNonceDict(nonce: nonce),
        ])
      }
    } else {
      payPalClient.tokenize(paypalRequest as! BTPayPalCheckoutRequest) { nonce, err in
        guard err == nil else {
          result([
            "error": [
              "message": String(describing: (err as? NSError)?.localizedDescription),
              "code": String(describing: (err as? NSError)?.code),
            ]
          ])
          return
        }
        result([
          "data": FlutterBraintreePluginHelper.buildPaymentNonceDict(nonce: nonce),
        ])
      }
    }
  }
  
  
  private func tokenizeVenmoAccount(_ call: FlutterMethodCall, apiClient: BTAPIClient, result: @escaping FlutterResult) {
    
    guard let venmoInfo = FlutterBraintreePluginHelper.dict(for: "request", in: call) else {
      FlutterBraintreePluginHelper.returnFlutterError(result: result, code: "-1", message: "Missing request parameters")
      return
    }
    
    var paymentMethod = BTVenmoPaymentMethodUsage.multiUse
    if let paymentMethodUsage = venmoInfo["paymentMethodUsage"] as? Int {
      paymentMethod = BTVenmoPaymentMethodUsage(rawValue: paymentMethodUsage) ?? .multiUse
    }
    
    let venmoRequest = BTVenmoRequest(paymentMethodUsage: paymentMethod)
    venmoRequest.profileID = venmoInfo["profileID"] as? String
    if let vault = venmoInfo["vault"] as? Bool {
      venmoRequest.vault = vault
    }
    if let isFinalAmount = venmoInfo["isFinalAmount"] as? Bool {
      venmoRequest.isFinalAmount = isFinalAmount
    }
    
    if let fallbackToWeb = venmoInfo["fallbackToWeb"] as? Bool {
      venmoRequest.fallbackToWeb = fallbackToWeb
    }
    
    venmoRequest.displayName = venmoInfo["displayName"] as? String
    if let collectCustomerBillingAddress = venmoInfo["collectCustomerBillingAddress"] as? Bool {
      venmoRequest.collectCustomerBillingAddress = collectCustomerBillingAddress
    }
    if let collectCustomerShippingAddress = venmoInfo["collectCustomerShippingAddress"] as? Bool {
      venmoRequest.collectCustomerShippingAddress = collectCustomerShippingAddress
    }
    venmoRequest.subTotalAmount = venmoInfo["subTotalAmount"] as? String
    venmoRequest.totalAmount = venmoInfo["totalAmount"] as? String
    venmoRequest.discountAmount = venmoInfo["discountAmount"] as? String
    venmoRequest.shippingAmount = venmoInfo["shippingAmount"] as? String
    venmoRequest.taxAmount = venmoInfo["taxAmount"] as? String
    if let lineItems = venmoInfo["lineItems"] as? [Any] {
      venmoRequest.lineItems = FlutterBraintreePluginHelper.makeVenmoItems(from: lineItems)
    }
    
    let venmoClient = BTVenmoClient(apiClient: apiClient)
    
    venmoClient.tokenize(venmoRequest, completion: { nonce, error in
      guard error == nil else {
        result([
          "error": [
            "message": String(describing: (error as? NSError)?.localizedDescription),
            "code": String(describing: (error as? NSError)?.code),
          ]
        ])
        return
      }
      result([
        "data": FlutterBraintreePluginHelper.buildPaymentNonceDict(nonce: nonce),
      ])
    })
  }
  
  
  private func deletePaymentMethodNonce(apiClient: BTAPIClient, nonce:String, result: @escaping FlutterResult) {
    
    let parameters = [
      "operationName": "DeletePaymentMethodFromSingleUseToken",
      "query": "mutation DeletePaymentMethodFromSingleUseToken($input: DeletePaymentMethodFromSingleUseTokenInput!) { deletePaymentMethodFromSingleUseToken(input: $input) { clientMutationId } }",
      "variables": [
        "input": [ "singleUseTokenId" : nonce ]
      ]
    ] as [String : Any]
    
    apiClient.post("", parameters: parameters, httpType: .graphQLAPI) { body, response, error in
      guard error == nil else {
        result([
          "error": [
            "code": String(describing: (error as? NSError)?.code),
            "message": error?.localizedDescription ?? "No vaulted payment methods",
          ]
        ])
        return
      }
      result(true)
    }
  }


  private func fetchPaymentMethodNonces(apiClient: BTAPIClient, defaultFirst:Bool, result: @escaping FlutterResult) {
    apiClient.fetchOrReturnRemoteConfiguration { config, error in
      guard error == nil else {
        result([
          "error": [
            "code": String(describing: (error as? NSError)?.code),
            "message": error?.localizedDescription ?? "No vaulted payment methods",
          ]
        ])
        return
      }
      apiClient.fetchPaymentMethodNonces(defaultFirst, completion: { methods, err in
        guard err == nil else {
          result([
            "error": [
              "code": String(describing: (err as? NSError)?.code),
              "message": err?.localizedDescription ?? "No vaulted payment methods",
            ]
          ])
          return
        }

        guard let methods = methods else {
          result([
            "error": [
              "code": "-1",
              "message": "No vaulted payment methods",
            ]
          ])
          return
        }
        result([
          "methods": methods.map({ nonce in
            return FlutterBraintreePluginHelper.buildPaymentNonceDict(nonce: nonce)
          }),
        ])
      })
    }
  }
}

File: android/src/main/kotlin/com/example/flutter_braintree_plugin/FlutterBraintreePlugin.kt
package com.example.flutter_braintree_plugin

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.braintreepayments.api.BraintreeClient
import com.braintreepayments.api.PayPalCheckoutRequest
import com.braintreepayments.api.PayPalClient
import com.braintreepayments.api.PayPalLineItem
import com.braintreepayments.api.PayPalPaymentIntent
import com.braintreepayments.api.PayPalRequest
import com.braintreepayments.api.PayPalVaultRequest
import com.braintreepayments.api.PostalAddress

import com.example.flutter_braintree_plugin.FlutterBraintreePluginHelper
import java.util.ArrayList

/** FlutterBraintreePlugin */
class FlutterBraintreePlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private var context: Context? = null

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_braintree_plugin")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    val authorization = FlutterBraintreePluginHelper.getAuthorization(call)
    if (authorization == null) {
      FlutterBraintreePluginHelper.returnAuthorizationMissingError(result)
      return
    }
    val apiClient = context?.let { BraintreeClient(it, authorization) }
    if (apiClient == null) {
      FlutterBraintreePluginHelper.returnAuthorizationMissingError(result)
      return
    }

    if (call.method == "tokenizePayPalAccount") {
      tokenizePayPalAccount(call, apiClient, result)
      return
    } else if (call.method == "tokenizeVenmoAccount") {
      tokenizeVenmoAccount(call, apiClient, result)
      return
    } else if (call.method == "fetchPaymentMethodNonces") {
      val defaultFirst = FlutterBraintreePluginHelper.bool("defaultFirst", call) ?: true
      fetchPaymentMethodNonces(apiClient, defaultFirst, result)
      return
    } else if (call.method == "deletePaymentMethodNonce") {
      val nonce = FlutterBraintreePluginHelper.string("payment_nonce", call)
      if (nonce == null) {
        result(mapOf(
          "error" to mapOf(
            "code" to "-1",
            "message" to "Payment method does not exist"
          )
        ))
        return
      }
      deletePaymentMethodNonce(apiClient, nonce, result)
      return
    }
    result.notImplemented()
  }

  private fun tokenizeVenmoAccount(call: MethodCall, apiClient: BTAPIClient, result: Result) {
    val driver = BTVenmoDriver(apiClient)

    val venmoInfo = FlutterBraintreePluginHelper.dict("request", call) ?: run {
      FlutterBraintreePluginHelper.returnFlutterError(result, "-1", "Missing request parameters")
      return
    }

    val venmoRequest = BTVenmoRequest()
    venmoRequest.profileID = venmoInfo["profileID"] as? String
    val vault = venmoInfo["vault"] as? Boolean
    vault?.let { venmoRequest.vault = it }
    val paymentMethodUsage = venmoInfo["paymentMethodUsage"] as? Int
    when (paymentMethodUsage) {
      1 -> venmoRequest.paymentMethodUsage = BTVenmoPaymentMethodUsage.unspecified
      2 -> venmoRequest.paymentMethodUsage = BTVenmoPaymentMethodUsage.singleUse
      else -> venmoRequest.paymentMethodUsage = BTVenmoPaymentMethodUsage.multiUse
    }
    venmoRequest.displayName = venmoInfo["displayName"] as? String
    val collectCustomerBillingAddress = venmoInfo["collectCustomerBillingAddress"] as? Boolean
    collectCustomerBillingAddress?.let { venmoRequest.collectCustomerBillingAddress = it }
    val collectCustomerShippingAddress = venmoInfo["collectCustomerShippingAddress"] as? Boolean
    collectCustomerShippingAddress?.let { venmoRequest.collectCustomerShippingAddress = it }
    venmoRequest.subTotalAmount = venmoInfo["subTotalAmount"] as? String
    venmoRequest.totalAmount = venmoInfo["totalAmount"] as? String
    venmoRequest.discountAmount = venmoInfo["discountAmount"] as? String
    venmoRequest.shippingAmount = venmoInfo["shippingAmount"] as? String
    venmoRequest.taxAmount = venmoInfo["taxAmount"] as? String
    val lineItems = venmoInfo["lineItems"] as? List<Any>
    lineItems?.let { venmoRequest.lineItems = FlutterBraintreePluginHelper.makeVenmoItems(it) }

    driver.tokenizeVenmoAccount(venmoRequest) { nonce, error in
      guard error == nil else {
        result(mapOf(
          "error" to mapOf(
            "message" to (error as? NSError)?.localizedDescription,
            "code" to (error as? NSError)?.code
          )
        ))
        return
      }
      result(mapOf(
        "data" to FlutterBraintreePluginHelper.buildPaymentNonceDict(nonce)
      ))
    }
  }

  private fun tokenizePayPalAccount(call: MethodCall, apiClient: BraintreeClient, result: Result) {
    val driver = PayPalClient(apiClient)

    val requestInfo = FlutterBraintreePluginHelper.dict("request", call) ?: run {
      FlutterBraintreePluginHelper.returnFlutterError(result, "-1", "Missing request parameters")
      return
    }

    val vault = requestInfo["vault"] as? Boolean ?: run {
      FlutterBraintreePluginHelper.returnFlutterError(result, "-1", "Missing request parameters")
      return
    }

    val paypalRequest:PayPalRequest
    if (vault) {
      val paypalVaultRequest = PayPalVaultRequest(true)
      val offerCredit = requestInfo["offerCredit"] as? Boolean
      if (offerCredit != null) {
        paypalVaultRequest.shouldOfferCredit = offerCredit
      }
      paypalRequest = paypalVaultRequest
    } else {
      val amount = requestInfo["amount"] as? String ?: "0"
      val paypalCheckoutRequest = PayPalCheckoutRequest(amount, true)
      paypalCheckoutRequest.currencyCode = requestInfo["currencyCode"] as? String
      val intent = requestInfo["intent"] as? Int
      when (intent) {
        1 -> paypalCheckoutRequest.intent = PayPalPaymentIntent.SALE
        2 -> paypalCheckoutRequest.intent = PayPalPaymentIntent.ORDER
        else -> paypalCheckoutRequest.intent = PayPalPaymentIntent.AUTHORIZE
      }
      val userAction = requestInfo["userAction"] as? Int
      when (userAction) {
        1 -> paypalCheckoutRequest.userAction = PayPalCheckoutRequest.USER_ACTION_COMMIT
        else -> paypalCheckoutRequest.userAction = PayPalCheckoutRequest.USER_ACTION_DEFAULT
      }
      val requestBillingAgreement = requestInfo["requestBillingAgreement"] as? Boolean
      requestBillingAgreement?.let { paypalCheckoutRequest.shouldRequestBillingAgreement = it }
      val offerPayLater = requestInfo["offerPayLater"] as? Boolean
      offerPayLater?.let { paypalCheckoutRequest.shouldOfferPayLater = it }
      paypalRequest = paypalCheckoutRequest
    }

    paypalRequest.displayName = requestInfo["displayName"] as? String
    val shippingAddressRequired = requestInfo["shippingAddressRequired"] as? Boolean
    shippingAddressRequired?.let { paypalRequest.isShippingAddressRequired = it }
    val billingAgreementDescription = requestInfo["billingAgreementDescription"] as? String
    billingAgreementDescription?.let { paypalRequest.billingAgreementDescription = it }
    val shippingAddressOverride = requestInfo["shippingAddressOverride"] as? Map<String, Any>
    shippingAddressOverride?.let {
      val address = PostalAddress()
        address.recipientName = it["recipientName"] as? String
        address.streetAddress = it["streetAddress"] as? String
        address.extendedAddress = it["extendedAddress"] as? String
        address.locality = it["locality"] as? String
        address.region = it["region"] as? String
        address.postalCode = it["postalCode"] as? String
        address.countryCodeAlpha2 = it["countryCodeAlpha2"] as? String
        paypalRequest.shippingAddressOverride = address
    }
    val shippingAddressEditable = requestInfo["shippingAddressEditable"] as? Boolean
    shippingAddressEditable?.let { paypalRequest.isShippingAddressEditable = it }
    val localeCode = requestInfo["localeCode"] as? String
    localeCode?.let { paypalRequest.localeCode = it }
    val merchantAccountID = requestInfo["merchantAccountID"] as? String
    merchantAccountID?.let { paypalRequest.merchantAccountId = it }
    val riskCorrelationId = requestInfo["riskCorrelationId"] as? String
    riskCorrelationId?.let { paypalRequest.riskCorrelationId = it }

    val landingPageType = requestInfo["landingPageType"] as? Int
    when (landingPageType) {
      1 -> paypalRequest.landingPageType = PayPalRequest.LANDING_PAGE_TYPE_LOGIN
      2 -> paypalRequest.landingPageType = PayPalRequest.LANDING_PAGE_TYPE_BILLING
      else -> paypalRequest.landingPageType = PayPalRequest.LANDING_PAGE_TYPE_LOGIN
    }

    val lineItems = requestInfo["lineItems"] as? List<Any>
    lineItems?.let {
      val items = FlutterBraintreePluginHelper.makePayPalItems(it)
      if (items != null) {
        paypalRequest.setLineItems(items)
      }
    }

    driver.tokenizePayPalAccount()
    driver.tokenizePayPalAccount(paypalRequest) { nonce, error in
      guard error == nil else {
        result(mapOf(
          "error" to mapOf(
            "message" to (error as? NSError)?.localizedDescription,
            "code" to (error as? NSError)?.code
          )
        ))
        return
      }
      result(mapOf(
        "data" to FlutterBraintreePluginHelper.buildPaymentNonceDict(nonce)
      ))
    }
  }

  private fun deletePaymentMethodNonce(apiClient: BTAPIClient, nonce:String, result: Result) {
    val parameters = mapOf(
      "operationName" to "DeletePaymentMethodFromSingleUseToken",
      "query" to "mutation DeletePaymentMethodFromSingleUseToken($input: DeletePaymentMethodFromSingleUseTokenInput!) { deletePaymentMethodFromSingleUseToken(input: $input) { clientMutationId } }",
      "variables" to mapOf(
        "input" to mapOf( "singleUseTokenId" to nonce )
      )
    )

    apiClient.post("", parameters, httpType: .graphQLAPI) { body, response, error in
      guard error == nil else {
        result(mapOf(
          "error" to mapOf(
            "code" to (error as? NSError)?.code.toString(),
            "message" to error?.localizedDescription ?: "No vaulted payment methods"
          )
        ))
        return
      }
      fetchPaymentMethodNonces(apiClient, true, result)
    }
  }


  private fun fetchPaymentMethodNonces(apiClient: BTAPIClient, defaultFirst:Boolean, result: Result) {
    apiClient.fetchOrReturnRemoteConfiguration { config, error in
      guard error == null else {
        result(mapOf(
          "error" to mapOf(
            "code" to (error as? NSError)?.code.toString(),
            "message" to error?.localizedDescription ?: "No vaulted payment methods"
          )
        ))
        return
      }
      apiClient.fetchPaymentMethodNonces(defaultFirst, completion: { methods, err in
        guard err == null else {
          result(mapOf(
            "error" to mapOf(
              "code" to (err as? NSError)?.code.toString(),
              "message" to err?.localizedDescription ?: "No vaulted payment methods"
            )
          ))
          return
        }

        guard let methods = methods else {
          result(mapOf(
            "error" to mapOf(
              "code" to "-1",
              "message" to "No vaulted payment methods"
            )
          ))
          return
        }
        result(mapOf(
          "methods" to methods.map({ nonce in
                  return FlutterBraintreePluginHelper.buildPaymentNonceDict(nonce: nonce)
          }),
        ))
      })
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    context = null
    channel.setMethodCallHandler(null)
  }
}
